### PR TITLE
Enable the `%%file` xmagic

### PR DESCRIPTION
**Pull request by faze-geek**
# Description

Towards #176

## Type of change

Please tick all options which are relevant.

- [ ] Bug fix
- [x] New feature
- [ ] Added/removed dependencies
- [ ] Required documentation updates

### CODE REVIEW

**Comment by faze-geek** [comment]
Let me know if I need to add more tests. I wrote them for whatever functionalities were supported in the documentation.

**Comment by anutosh491** [comment]
Please make the following changes 

1) In cmakelists.txt : Move oss header & source to the not emscripten case.
```
set(XEUS_CPP_HEADERS
    ....
    #src/xmagics/os.hpp
)

set(XEUS_CPP_SRC
    ....
    src/xinterpreter.cpp
    src/xmagics/os.cpp
    ....
)

if(NOT EMSCRIPTEN)
    list(APPEND XEUS_CPP_SRC
        src/xmagics/xassist.hpp
        src/xmagics/xassist.cpp
    )
endif()
```

2) in `src/xinterpreter.cpp` : Do the same as above.

```
#include "xmagics/os.hpp"
#ifndef EMSCRIPTEN
#include "xmagics/xassist.hpp"
#endif
```

**Comment by anutosh491** [comment]
Apart from that. The tests look decent to me. Maybe @mcbarton could also review this for us !

**Comment by vgvassilev** [comment]
@mcbarton, do we not have code coverage bots checking the test suite coverage for xeus-cpp? I remember we had that but maybe I am confusing myself...

**Comment by mcbarton** [comment]
> @mcbarton, do we not have code coverage bots checking the test suite coverage for xeus-cpp? I remember we had that but maybe I am confusing myself...

Isn't the test coverage being shown in this message? https://github.com/compiler-research/xeus-cpp/pull/181#issuecomment-2482751686

File: CMakeLists.txt [hunk 1]
```diff
@@ -186,6 +184,8 @@ if(NOT EMSCRIPTEN)
     list(APPEND XEUS_CPP_SRC
         src/xmagics/xassist.hpp
         src/xmagics/xassist.cpp
+        src/xmagics/os.hpp
```
Comment (vgvassilev) [inline CMakeLists.txt:187]: Do we need to list the local `hpp` files?

Comment (anutosh491) [reply CMakeLists.txt]: Guessing we can do without them. Trying it out.

Comment (anutosh491) [reply CMakeLists.txt]: The tests pass, so I think we can do without them

File: CMakeLists.txt [hunk 2]
```diff
@@ -169,23 +169,23 @@ set(XEUS_CPP_HEADERS
...
     src/xutils.cpp
 )
 
 if(NOT EMSCRIPTEN)
     list(APPEND XEUS_CPP_SRC
-        src/xmagics/xassist.hpp
+        #src/xmagics/xassist.hpp
         src/xmagics/xassist.cpp
+        #src/xmagics/os.hpp
+        src/xmagics/os.cpp
```
Comment (vgvassilev) [inline CMakeLists.txt:188]: Can we remove the commented lines?

Comment (anutosh491) [reply CMakeLists.txt:186]: Done.

**Comment by anutosh491** [comment]
I am guessing this is ready ?

**Comment by vgvassilev** [comment]
The suggested by clang-format changes look good.

**Comment by faze-geek** [comment]
> The suggested by clang-format changes look good.
Done !

**Comment by anutosh491** [comment]
I see a duplication of tests here. We already had tests for the writefile operation.

https://github.com/compiler-research/xeus-cpp/blob/ca8c42e95b6b1aa5a84f31d6285af83e0de5cf04/test/test_interpreter.cpp#L527-L590

Can someone confirm this for me. The tests look similar though this looks better framed. Maybe @kr-2003 would like to help (maybe get rid of the first set of tests)

**Comment by anutosh491** [comment]
> The only difference I can see is "removing the txt file after every subtest".

Not worth it. Let's just keep the better test

**Pull request merged**